src/json/next/serialization.cr

```diff
@@ -0,0 +1,233 @@
+module JSON
+  annotation Field
+  end
+
+  module Serializable
+    annotation Options
+    end
+
+    macro included
+      # Define a `new` directly in the included type,
+      # so it overloads well with other possible initializes
+
+      def self.new(pull : ::JSON::PullParser)
+        instance = allocate
```

Comment:
Does replacing this method body with just `previous_def` really break this hack?

Reply:
not compiled, `there is no previous definition of 'new'`

Reply:
Maybe we should have a new keyword to do `previous_def_or_super` kind of behavior? (it's not the first time I needed this thing iirc)

Reply:
I don't understand why this snippet doesn't work but this one does then: https://carc.in/#/r/44bt

```cr
class Foo
  def initialize(@foo : String)
    {{ @type }}
  end
  
  macro finished
    def self.new(foo : String)
      if rand > 0.5
        previous_def(foo)
      else
        Foo.new("bar")
      end
    end
  end
end


p (1..10).map { Foo.new("foo") }   
```

Reply:
My concern is when we refactor `.new` to change something and forget all the `allocate/initialize/finalizer` code that needs to be updated. And all this for a bug: we shouldn't need to do that.

It's okay to merge if we really can't figure it out but i'd like to take a good long look at fixing it if we can.

Reply:
Actually, for `JSON::Serializable` this `new` is not needed. `initialize` works well. I just did it like that for consistency with `YAML::Serializable`. So you can change it if you want.